doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+  std::cout << "subscribed!\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, onWaitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void onWaitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, onWaitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED);
+}
+
+int main() {
+  myCallbackReactal.attachEvent(mySubscriber, iox::popo::SubscriberEvent::SUBSCRIBED, onWaitForSubscription);
+}
+```
+
+### Solution
+#### Condition Variable
+
+  - Alternative names: `EventVariable`, `EventSignaler`, `EventWaiter`
+  
+  - **Problem:** The ReactAl would like to know by whom it was triggered. The WaitSet has a 
+                  big list of callbacks where it can ask the object if it triggered the WaitSet. This has 
+                  certain disadvantages.
+              - Cache misses and low performance 
+              - The object could lie due to a bug.
+              - Implementation overhead since we have to manage callbacks.
+
+  - **Solution:** The condition variable tracks by whom it was triggered. To realize this every signaler
+      gets provided with an id. When the condition variable is notified a corresponding bool array entry is set to true 
+      to trace the trigger origin.
+
+  - Create from `ConditionVariableWithOriginTracing` (yeah we need a better name) has `ConditionVariableData` 
+    as a member and adds: `std::atomic_bool m_triggeredBy[MAX_CALLBACKS];`
+
+    **Reason:**
+      - The reactal will iterate through this array to find out by which event it was triggered 
+      - It is cache local, since it is one piece of contiguous memory which can be put directly 
+        into the CPU cache when it is being iterated. Therefore it is very performant.
+      - WaitSet approach are callbacks which have an overhead implementation 
+        and performance wise.
+        * implementation, since we have to manage set and reset all the callbacks 
+        * performance, when we call a callback with every iteration, we have to first load the 
+            callback into the cpu cache (generates cache misses for all the other callback 
+            elements in the vector which is being iterated). In the callback a method in an 
+            object is called (the object has to be loaded again into the cpu cache, cache miss for 
+            every member which is not used by the method).
+
+    **Potential optimization:**
+      - Create an additional bool member `m_wasTriggered` which is true when at least 
```

Comment:
Should be a flag in the array (e.g. first or last one) itself for cache efficiency.